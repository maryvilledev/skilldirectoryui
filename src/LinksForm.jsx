import React from 'react';
import axios from 'axios';
import 'react-select/dist/react-select.css';

var Select = require('react-select');

var linkTypes = [
  { value: 'blog', label: 'Blog' },
  { value: 'tutorial', label: 'Tutorial'},
  { value: 'webpage', label: 'Webpage' }
];

class LinksForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      link_name: "",
      link_url: "",
      link_type: "",
    }
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(key, value) {
    this.setState({ [key]: value });
  }

  onSubmit(ev) {
    ev.preventDefault();
    const skill_id = this.props.skill_id;
    const link_name = this.state.link_name;
    const link_url = this.state.link_url;
    const link_type = this.state.link_type;

    // Check validity of URL before sending POST request
    if(!isValidURL(link_url)) {
      alert('Please enter a valid URL.');
      return;
    }

    // Post form data to API endpoint
    axios.post(this.props.api + '/links/', {
      name: link_name,
      url: link_url,
      skill_id: skill_id,
      link_type: link_type
    })
    .then(function (response) {
      console.log(response);
      this.props.closeModal();
    }.bind(this))
    .catch(err => {
      console.log('caught an error', err);
    });
  }

  render() {
    const onLinkNameChange = ev => this.onChange('link_name', ev.target.value);
    const onLinkURLChange = ev => this.onChange('link_url', ev.target.value);
    const onLinkTypeChange = ev => this.onChange('link_type', ev.value);
    return (
      <form onSubmit={ev => this.onSubmit(ev)}>
        <div>
          Name: <input name='link_name'
                       type='text'
                       value={this.state.link_name}
                       onChange={onLinkNameChange} />
        </div>
        <div>
          URL: <input name='link_url'
                      type='text'
                      value={this.state.link_url}
                      onChange={onLinkURLChange} />
        </div>
        <div>
          Type: <Select name='link_type'
                        value={this.state.link_type}
                        onChange={onLinkTypeChange}
                        options={linkTypes} />
        </div>
        <button type="submit">Save</button>
      </form>
    )
  }
}

/* eslint-disable no-useless-escape */
/* http://stackoverflow.com/a/30970319 */
function isValidURL(url) {
    var res = url.match(/(http(s)?:\/\/.)(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g);
    if(res == null)
        return false;
    else
        return true;
}

export default LinksForm
module.exports.isValidURL = isValidURL
