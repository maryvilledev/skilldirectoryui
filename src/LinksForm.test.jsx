import React from 'react';
import ReactDOM from 'react-dom';
import LinksForm from './LinksForm';
import { isValidURL } from './LinksForm';
import { shallow, mount } from 'enzyme';

describe('<LinksForm />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<LinksForm />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<LinksForm />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('changes state when input to "Name" field changes', () => {
    const wrapper = mount(<LinksForm />);
    // process.stdout.write(wrapper.debug());
    // process.stdout.write(JSON.stringify(wrapper.state()));
    expect(wrapper.state().link_name).toBe("");
    const event = {target: {value: "New Value"}};
    wrapper.find({ name: "link_name" }).simulate("change", event);
    expect(wrapper.state().link_name).toBe("New Value");
    // process.stdout.write(JSON.stringify(wrapper.state()));
  });

  it('changes state when input to "URL" field changes', () => {
    const wrapper = mount(<LinksForm />);
    expect(wrapper.state().link_url).toBe("");
    const event = {target: {value: "New Value"}};
    wrapper.find({ name: "link_url" }).simulate("change", event);
    expect(wrapper.state().link_url).toBe("New Value");
  });

// TODO: Find different way to test this... The below test fails because
// the "Select" element we are using from 'react-select' doesn't respond
// as expected to the simulated "change" event.
  // it('changes state when selected option in "Type" field changes', () => {
  //   const wrapper = mount(<LinksForm />);
  //   expect(wrapper.state().link_type).toBe("");
  //   const event = {target: {value: "New Value"}};
  //   wrapper.find("Select").simulate("change", event);
  //   expect(wrapper.state().link_type).toBe("New Value");
  // });
});

test('isValidURL() successfully detects valid/invalid URL forms', () => {
  expect(isValidURL("http://www.google.com")).toBe(true);
  expect(isValidURL("https://www.github.com")).toBe(true);
  expect(isValidURL("www.google.com")).toBe(false);
  expect(isValidURL("definitely-not-a-valid-url")).toBe(false);
});