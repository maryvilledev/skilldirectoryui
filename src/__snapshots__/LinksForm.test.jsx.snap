exports[`<LinksForm /> matches the stored snapshot 1`] = `
<form
  onSubmit={[Function]}>
  <div>
    Name: 
    <input
      name="link_name"
      onChange={[Function]}
      type="text"
      value="" />
  </div>
  <div>
    URL: 
    <input
      name="link_url"
      onChange={[Function]}
      type="text"
      value="" />
  </div>
  <div>
    Type: 
    <Select
      addLabelText="Add \"{label}\"?"
      arrowRenderer={[Function]}
      autosize={true}
      backspaceRemoves={true}
      backspaceToRemoveMessage="Press backspace to remove {label}"
      clearAllText="Clear all"
      clearValueText="Clear value"
      clearable={true}
      delimiter=","
      disabled={false}
      escapeClearsValue={true}
      filterOptions={[Function]}
      ignoreAccents={true}
      ignoreCase={true}
      inputProps={Object {}}
      isLoading={false}
      joinValues={false}
      labelKey="label"
      matchPos="any"
      matchProp="any"
      menuBuffer={0}
      menuRenderer={[Function]}
      multi={false}
      name="link_type"
      noResultsText="No results found"
      onBlurResetsInput={true}
      onChange={[Function]}
      onCloseResetsInput={true}
      openAfterFocus={false}
      optionComponent={[Function]}
      options={
        Array [
          Object {
            "label": "Blog",
            "value": "blog",
          },
          Object {
            "label": "Tutorial",
            "value": "tutorial",
          },
          Object {
            "label": "Webpage",
            "value": "webpage",
          },
        ]
      }
      pageSize={5}
      placeholder="Select..."
      required={false}
      scrollMenuIntoView={true}
      searchable={true}
      simpleValue={false}
      tabSelectsValue={true}
      value=""
      valueComponent={[Function]}
      valueKey="value" />
  </div>
  <button
    type="submit">
    Save
  </button>
</form>
`;
